xmtp_mls/src/storage/encrypted_store/v2_conversation.rs

```diff
@@ -0,0 +1,155 @@
+//! The V2 Conversation database table. Stored information surrounding v2 conversations.
+
+use crate::{impl_store, storage::StorageError};
+
+use super::Sqlite;
+use super::{
+    db_connection::DbConnection,
+    schema::consent_records::{self, dsl},
+};
+use diesel::{
+    backend::Backend,
+    deserialize::{self, FromSql, FromSqlRow},
+    expression::AsExpression,
+    prelude::*,
+    serialize::{self, IsNull, Output, ToSql},
+    sql_types::Integer,
+    upsert::excluded,
+};
+use serde::{Deserialize, Serialize};
+
+/// StoredV2Conversation holds a serialized ConsentRecord
+#[derive(Insertable, Queryable, Debug, Clone, PartialEq, Eq)]
+#[diesel(table_name = v2_conversations)]
+#[diesel(primary_key(topic))]
+pub struct StoredV2Conversation {
+    pub topic: String,
+    pub peer_address: String,
+    pub envelope_bytes: Vec<u8>,
+    pub created_at_ns: i64,
+}
+
+impl StoredV2Conversation {
+    pub fn new(topic: String, peer_address: String, envelope_bytes: Vec<u8>, created_at_ns: i64) -> Self {
+        Self {
+            topic,
+            peer_address,
+            envelope_bytes,
+            created_at_ns
+        }
+    }
+}
+
+impl_store!(StoredV2Conversation, v2_conversations);
+
+impl DbConnection {
+    /// Returns a v2_conversation
+    pub fn get_v2_conversation(
+        &self,
+        peer_address: String,
+    ) -> Result<Option<StoredV2Conversation>, StorageError> {
+        Ok(self.raw_query(|conn| -> diesel::QueryResult<_> {
+            dsl::v2_conversations
+                .filter(dsl::peer_address.eq(peer_address))
+                .first(conn)
+                .optional()
+        })?)
+    }
+
+    /// Returns all the v2_conversations
+    pub fn get_v2_conversations(
+        &self,
+    ) -> Result<Vec<StoredV2Conversation>, StorageError> {
+        Ok(self.raw_query(|conn| -> diesel::QueryResult<_> {
+            dsl::v2_conversations.load(conn)
+        })?)
+    }
+
+    /// Insert v2_conversations
+    pub fn insert_or_replace_v2_conversation(&self, v2_conversation: StoredV2Conversation) -> Result<StoredV2Conversation, StorageError> {
+        tracing::info!("Trying to insert v2 conversation");
+        let stored_v2_conversation = self.raw_query(|conn| {
+            let maybe_inserted_conversation: Option<StoredV2Conversation> = diesel::insert_into(dsl::v2_conversations)
```

Comment:
If this is using `on_conflict_do_nothing` then it's actually an "insert or ignore" not an "insert or replace"

Reply:
There is also a helpful Diesel helper function `insert_or_ignore_into`